spec/models/manageiq/providers/nuage/vsd_client/vcr_specs/vsd_client_spec.rb

```diff
@@ -0,0 +1,596 @@
+require 'json'
+describe ManageIQ::Providers::Nuage::NetworkManager::VsdClient do
+  include Vmdb::Logging
+  before(:each) do
+    @userid = Rails.application.secrets.nuage_network.try(:[], 'userid') || 'NUAGE_USER_ID'
+    @password = Rails.application.secrets.nuage_network.try(:[], 'password') || 'NUAGE_PASSWORD'
+    @hostname = Rails.application.secrets.nuage_network.try(:[], 'host') || 'nuagenetworkhost'
+
+    # Ensure that VCR will obfuscate the basic auth
+    VCR.configure do |c|
+      # workaround for escaping host
+      c.before_playback do |interaction|
+        interaction.filter!(CGI.escape(@hostname), @hostname)
+        interaction.filter!(CGI.escape('NUAGE_NETWORK_HOST'), 'nuagenetworkhost')
+      end
+      c.filter_sensitive_data('NUAGE_NETWORK_AUTHORIZATION') { Base64.encode64("#{@userid}:#{@password}").chomp }
+    end
+  end
+
+  it "should return valid non empty json response for enterprises" do
+    enterprises = nil
+    VCR.use_cassette(described_class.parent.name.underscore + '/vsd_client/enterprises', :record => :new_episodes) do
+      vsd_client = described_class.new("https://#{@hostname}:8443/nuage/api/v5_0", @userid, @password)
+      enterprises = vsd_client.get_enterprises
+    end
+
+    assert_object_not_empty(enterprises)
+    assert_enterprises_parameters(enterprises)
+  end
+
+  it "should return empty response for enterprises" do
+    enterprises = nil
+    VCR.use_cassette(described_class.parent.name.underscore + '/vsd_client/enterprises_empty', :record => :new_episodes) do
+      vsd_client = described_class.new("https://#{@hostname}:8443/nuage/api/v5_0", @userid, @password)
+      enterprises = vsd_client.get_enterprises
+    end
+
+    expect(enterprises).to be_nil
+  end
+
+  it "should return valid non empty json response for domains" do
+    domains = nil
+    VCR.use_cassette(described_class.parent.name.underscore + '/vsd_client/domains', :record => :new_episodes) do
+      vsd_client = described_class.new("https://#{@hostname}:8443/nuage/api/v5_0", @userid, @password)
+      domains = vsd_client.get_domains
+    end
+
+    assert_object_not_empty(domains)
+    assert_domains_parameters(domains)
+  end
+
+  it "should return empty response for domains" do
```

Comment:
I am wondering - what does this do? You are calling invalid endpoint (`.../domains_empty`).

Reply:
No, i am using a VCR cassette named `domains_empty`. I have manually fixed the response in that cassette file to return UTF-8 encoded empty string, that is what the `vsd_client` expects in case the 'backend' has no data to return. This is how we cover the part of `vsd_client` methods which handle empty responses.